suite-sparse build: cannot find gmp.h

Installing `suite-sparse@5.8.1` using:
* `spack@develop`
* commit d1493fb3d86653d8b11aeacd3e24af94f50ec03c from `Thu Nov 12 06:57:44 2020 -0800`
* `Ubuntu 18.04` for `x86_64`

Concrete spec: [suite-sparse.spec.yaml.txt](https://github.com/spack/spack/files/5531743/suite-sparse.spec.yaml.txt)
Full build log: [suite-sparse-build-out.txt](https://github.com/spack/spack/files/5531809/suite-sparse-build-out.txt)

```
$> spack install -f ./suite-sparse.spec.yaml
...
==> Installing suite-sparse
==> Using cached archive: /opt/spack/var/spack/cache/_source-cache/archive/06/06726e471fbaa55f792578f9b4ab282ea9d008cf39ddcc3b42b73400acddef40.tar.gz
==> suite-sparse: Executing phase: 'install'
==> Error: ProcessError: Command exited with status 2:
    'make' '-j16' 'library' 'AUTOCC=no' 'CUDA=no' 'CUDA_PATH=' 'CFOPENMP=' 'CFLAGS=-O3 -fPIC' 'FFLAGS=-fPIC' 'F77FLAGS=-fPIC' 'MY_METIS_LIB=-L/opt/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.5.0/metis-5.1.0-edcrft4l4szpqji2kowdkzq2ofueh3dr/lib -lmetis' 'MY_METIS_INC=/opt/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.5.0/metis-5.1.0-edcrft4l4szpqji2kowdkzq2ofueh3dr/include' 'BLAS=-L/opt/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.5.0/openblas-0.3.12-te35tjxctgjpeeuk357xghoinvwf2k7u/lib -lopenblas' 'LAPACK=-L/opt/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.5.0/openblas-0.3.12-te35tjxctgjpeeuk357xghoinvwf2k7u/lib -lopenblas' 'CFLAGS+=-fno-common -fexceptions' 'CMAKE_OPTIONS=-DCMAKE_INSTALL_PREFIX=/opt/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.5.0/suite-sparse-5.8.1-w26hi7moz73y7dsz5rresxfyzsra2tcf -DCMAKE_LIBRARY_PATH=/opt/spack/opt/spack/linux-ubuntu18.04-x86_64/gcc-7.5.0/suite-sparse-5.8.1-w26hi7moz73y7dsz5rresxfyzsra2tcf/lib'

35 errors found in build log:
     5188    /opt/spack/lib/spack/env/gcc/gcc -O3 -fPIC -fno-common -fexceptions   -O3 -fexceptions -fPIC  -I../Include -I../../COLAMD/Include -I../../AMD/I
             nclude -I../../SuiteSparse_config  -c ../Source/slip_back_sub.c
     5189    /opt/spack/lib/spack/env/gcc/gcc -O3 -fPIC -fno-common -fexceptions   -O3 -fexceptions -fPIC  -I../Include -I../../COLAMD/Include -I../../AMD/I
             nclude -I../../SuiteSparse_config  -c ../Source/slip_cumsum.c
     5190    /opt/spack/lib/spack/env/gcc/gcc -O3 -fPIC -fno-common -fexceptions   -O3 -fexceptions -fPIC  -I../Include -I../../COLAMD/Include -I../../AMD/I
             nclude -I../../SuiteSparse_config  -c ../Source/slip_get_pivot.c
     5191    /opt/spack/lib/spack/env/gcc/gcc -O3 -fPIC -fno-common -fexceptions   -O3 -fexceptions -fPIC  -I../Include -I../../COLAMD/Include -I../../AMD/I
             nclude -I../../SuiteSparse_config  -c ../Source/SLIP_malloc.c
     5192    In file included from ../Source/slip_internal.h:106:0,
     5193                     from ../Source/SLIP_free.c:13:
  >> 5194    ../Include/SLIP_LU.h:153:10: fatal error: gmp.h: No such file or directory
     5195     #include <gmp.h>
     5196              ^~~~~~~
     5197    compilation terminated.
     5198    Makefile:46: recipe for target 'SLIP_free.o' failed
  >> 5199    make[3]: *** [SLIP_free.o] Error 1
     5200    make[3]: *** Waiting for unfinished jobs....
     5201    In file included from ../Source/slip_internal.h:106:0,
     5202                     from ../Source/slip_create_mpq_array.c:15:
  >> 5203    ../Include/SLIP_LU.h:153:10: fatal error: gmp.h: No such file or directory
     5204     #include <gmp.h>
     5205              ^~~~~~~
     5206    compilation terminated.
     5207    Makefile:46: recipe for target 'slip_create_mpq_array.o' failed
  >> 5208    make[3]: *** [slip_create_mpq_array.o] Error 1
     5209    In file included from ../Source/slip_internal.h:106:0,
     5210                     from ../Source/slip_matrix_mul.c:17:
  >> 5211    ../Include/SLIP_LU.h:153:10: fatal error: gmp.h: No such file or directory
     5212     #include <gmp.h>
     5213              ^~~~~~~
     5214    compilation terminated.
     5215    In file included from ../Source/slip_internal.h:106:0,
     5216                     from ../Source/slip_get_largest_pivot.c:23:
  >> 5217    ../Include/SLIP_LU.h:153:10: fatal error: gmp.h: No such file or directory
     5218     #include <gmp.h>
     5219              ^~~~~~~
     5220    compilation terminated.
     5221    Makefile:46: recipe for target 'slip_matrix_mul.o' failed
  >> 5222    make[3]: *** [slip_matrix_mul.o] Error 1
     5223    In file included from ../Source/slip_internal.h:106:0,
     5224                     from ../Source/slip_back_sub.c:22:
  >> 5225    ../Include/SLIP_LU.h:153:10: fatal error: gmp.h: No such file or directory
     5226     #include <gmp.h>
     5227              ^~~~~~~
     5228    Makefile:46: recipe for target 'slip_get_largest_pivot.o' failed
  >> 5229    make[3]: *** [slip_get_largest_pivot.o] Error 1
     5230    compilation terminated.
     5231    Makefile:46: recipe for target 'slip_back_sub.o' failed
  >> 5232    make[3]: *** [slip_back_sub.o] Error 1
     5233    In file included from ../Source/slip_internal.h:106:0,
     5234                     from ../Source/SLIP_realloc.c:11:
  >> 5235    ../Include/SLIP_LU.h:153:10: fatal error: gmp.h: No such file or directory
     5236     #include <gmp.h>
     5237              ^~~~~~~
     5238    compilation terminated.
     5239    Makefile:46: recipe for target 'SLIP_realloc.o' failed
  >> 5240    make[3]: *** [SLIP_realloc.o] Error 1
     5241    In file included from ../Source/slip_internal.h:106:0,
     5242                     from ../Source/SLIP_LU_solve.c:51:
  >> 5243    ../Include/SLIP_LU.h:153:10: fatal error: gmp.h: No such file or directory
     5244     #include <gmp.h>
     5245              ^~~~~~~
     5246    compilation terminated.
     5247    In file included from ../Source/slip_internal.h:106:0,
     5248                     from ../Source/SLIP_backslash.c:45:
  >> 5249    ../Include/SLIP_LU.h:153:10: fatal error: gmp.h: No such file or directory
     5250     #include <gmp.h>
     5251              ^~~~~~~
     5252    compilation terminated.
     5253    In file included from ../Source/slip_internal.h:106:0,
     5254                     from ../Source/slip_matrix_div.c:26:
  >> 5255    ../Include/SLIP_LU.h:153:10: fatal error: gmp.h: No such file or directory
     5256     #include <gmp.h>
     5257              ^~~~~~~
     5258    compilation terminated.
     5259    Makefile:46: recipe for target 'SLIP_LU_solve.o' failed
  >> 5260    make[3]: *** [SLIP_LU_solve.o] Error 1
     5261    Makefile:46: recipe for target 'SLIP_backslash.o' failed
  >> 5262    make[3]: *** [SLIP_backslash.o] Error 1
     5263    Makefile:46: recipe for target 'slip_matrix_div.o' failed
  >> 5264    make[3]: *** [slip_matrix_div.o] Error 1
     5265    In file included from ../Source/slip_internal.h:106:0,
     5266                     from ../Source/SLIP_LU_factorize.c:40:
  >> 5267    ../Include/SLIP_LU.h:153:10: fatal error: gmp.h: No such file or directory
     5268     #include <gmp.h>
     5269              ^~~~~~~
     5270    compilation terminated.
     5271    In file included from ../Source/slip_internal.h:106:0,
     5272                     from ../Source/slip_ref_triangular_solve.c:70:
  >> 5273    ../Include/SLIP_LU.h:153:10: fatal error: gmp.h: No such file or directory
     5274     #include <gmp.h>
     5275              ^~~~~~~
     5276    compilation terminated.
     5277    Makefile:46: recipe for target 'SLIP_LU_factorize.o' failed
  >> 5278    make[3]: *** [SLIP_LU_factorize.o] Error 1
     5279    Makefile:46: recipe for target 'slip_ref_triangular_solve.o' failed
  >> 5280    make[3]: *** [slip_ref_triangular_solve.o] Error 1
     5281    In file included from ../Source/slip_internal.h:106:0,
     5282                     from ../Source/slip_get_nonzero_pivot.c:20:
  >> 5283    ../Include/SLIP_LU.h:153:10: fatal error: gmp.h: No such file or directory
     5284     #include <gmp.h>
     5285              ^~~~~~~
     5286    compilation terminated.
     5287    Makefile:46: recipe for target 'slip_get_nonzero_pivot.o' failed
  >> 5288    make[3]: *** [slip_get_nonzero_pivot.o] Error 1
     5289    In file included from ../Source/slip_internal.h:106:0,
     5290                     from ../Source/slip_cumsum.c:15:
  >> 5291    ../Include/SLIP_LU.h:153:10: fatal error: gmp.h: No such file or directory
     5292     #include <gmp.h>
     5293              ^~~~~~~
     5294    compilation terminated.
     5295    In file included from ../Source/slip_internal.h:106:0,
     5296                     from ../Source/slip_get_pivot.c:31:
  >> 5297    ../Include/SLIP_LU.h:153:10: fatal error: gmp.h: No such file or directory
     5298     #include <gmp.h>
     5299              ^~~~~~~
     5300    compilation terminated.
     5301    Makefile:46: recipe for target 'slip_cumsum.o' failed
  >> 5302    make[3]: *** [slip_cumsum.o] Error 1
     5303    In file included from ../Source/slip_internal.h:106:0,
     5304                     from ../Source/slip_create_mpz_array.c:15:
  >> 5305    ../Include/SLIP_LU.h:153:10: fatal error: gmp.h: No such file or directory
     5306     #include <gmp.h>
     5307              ^~~~~~~
     5308    Makefile:46: recipe for target 'slip_get_pivot.o' failed
     5309    compilation terminated.
  >> 5310    make[3]: *** [slip_get_pivot.o] Error 1
     5311    Makefile:46: recipe for target 'slip_create_mpz_array.o' failed
  >> 5312    make[3]: *** [slip_create_mpz_array.o] Error 1
     5313    In file included from ../Source/slip_internal.h:106:0,
     5314                     from ../Source/SLIP_malloc.c:13:
  >> 5315    ../Include/SLIP_LU.h:153:10: fatal error: gmp.h: No such file or directory
     5316     #include <gmp.h>
     5317              ^~~~~~~
     5318    compilation terminated.
     5319    Makefile:46: recipe for target 'SLIP_malloc.o' failed
  >> 5320    make[3]: *** [SLIP_malloc.o] Error 1
     5321    make[3]: Leaving directory '/tmp/root/spack-stage/spack-stage-suite-sparse-5.8.1-w26hi7moz73y7dsz5rresxfyzsra2tcf/spack-src/SLIP_LU/Lib'
     5322    Makefile:53: recipe for target 'library' failed
  >> 5323    make[2]: *** [library] Error 2
     5324    make[2]: Leaving directory '/tmp/root/spack-stage/spack-stage-suite-sparse-5.8.1-w26hi7moz73y7dsz5rresxfyzsra2tcf/spack-src/SLIP_LU/Lib'
     5325    Makefile:32: recipe for target 'library' failed
  >> 5326    make[1]: *** [library] Error 2
     5327    make[1]: Leaving directory '/tmp/root/spack-stage/spack-stage-suite-sparse-5.8.1-w26hi7moz73y7dsz5rresxfyzsra2tcf/spack-src/SLIP_LU'
     5328    Makefile:128: recipe for target 'library' failed
  >> 5329    make: *** [library] Error 2
```

Adding a `depends_on('gmp', type='build')` does not fix this.

@siko1056 @jppelteret @adamjstewart @t-karatsu @mathsen

--- a/var/spack/repos/builtin/packages/suite-sparse/package.py
+++ b/var/spack/repos/builtin/packages/suite-sparse/package.py
@@ -35,6 +35,8 @@ class SuiteSparse(Package):
     variant('cuda', default=False, description='Build with CUDA')
     variant('openmp', default=False, description='Build with OpenMP')
 
+    depends_on('mpfr', type=('build', 'link'))
+    depends_on('gmp', type=('build', 'link'))
     depends_on('blas')
     depends_on('lapack')
     depends_on('m4', type='build', when='@5.0.0:')

